# ELVTR Data Science Main Project

## Deliverables

### 01 Git Repository

Include all project code with a README file containing a high-level project description.

Example README guide: [Make a README](link-to-readme-guide)

### 02 Report

* Methodology, approach, and model selection rationale.
* Advantages and limitations of the chosen model.
* Architecture of the final solution.
* Considerations on deployment and scalability of the solution - i.e., how will the model be used in BAU by the business?
* Estimated impact/ROI of the project.

# Data Science in Finance: Lending Club Loan Analysis

## Project Overview

Lending Club has tasked us with preparing a loan application dataset for analysis and predictive modeling. 

The key tasks include data cleaning, exploratory data analysis, and building a predictive model for loan classification. An optional component involves building a real-time scoring application.

**Project Objectives**:
- Clean and preprocess the data.
- Perform exploratory data analysis (EDA) to gain insights.
- Develop a predictive model for loan application approval.
- (Optional) Build a real-time scoring application.

**Dataset Description**:
The dataset consists of loan application records, including various financial metrics and the application status. The data dictionary is provided for understanding the attributes.

**Dataset Path**:
- CSV: `data/1-raw/lending-club-2007-2020Q3/Loan_status_2007-2020Q3-100ksample.csv`
- Data Dictionary: `data/1-raw/lending-club-2007-2020Q3/LCDataDictionary.xlsx`

In [2]:
pwd

'c:\\Users\\kiera\\OneDrive\\Documents\\GitHub\\dsif-git-main-project\\elvtr_main_project\\notebooks'

## Importing Required Libraries


In [3]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# Set display options for pandas
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', 100)

## Load Data Set

In [4]:
# Load the data xlsx file as a dataframe
df = pd.read_csv("c:\\Users\\kiera\\OneDrive\\Documents\\GitHub\\dsif-git-main-project\\elvtr_main_project\\data\\raw\\Loan_status_2007-2020Q3\\Loan_status_2007-2020Q3-100k-Full-Data.csv")

df.shape

(100000, 143)

In [5]:
# Load the data dictionary CSV file as a dataframe
df_data_dict = pd.read_excel("c:\\Users\\kiera\\OneDrive\\Documents\\GitHub\\dsif-git-main-project\\elvtr_main_project\\data\\raw\\Loan_status_2007-2020Q3\\LCDataDictionary.xlsx")

df_data_dict.shape

(153, 2)

In [6]:
#load the employtment mapping CSV file as a dataframe
df_emp_title = pd.read_csv("c:\\Users\\kiera\\OneDrive\\Documents\\GitHub\\dsif-git-main-project\\elvtr_main_project\\data\\raw\\emp_title_mapping.csv")

df_emp_title.shape

(91, 2)

## Display Basic Data Discovery

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Columns: 143 entries, Unnamed: 0.1 to debt_settlement_flag
dtypes: float64(91), int64(18), object(34)
memory usage: 109.1+ MB


Our data contains 143 columns and 99999 rows of data. It is comprised of numerical (float, int) and categorical data (object)

In [8]:
df.head()

Unnamed: 0.1  Unnamed: 0         id  loan_amnt  funded_amnt  \
0       1746494       25450  167338079       4000         4000   
1       2370717       81861   71016917      24000        24000   
2       2264870      397109   39589826       5000         5000   
3        595422       15492  134798709      24000        24000   
4        562657       90591  127097355      14000        14000   

   funded_amnt_inv        term int_rate  installment grade sub_grade  \
0           4000.0   36 months   13.08%       134.93     B        B5   
1          24000.0   60 months    9.16%       500.07     B        B2   
2           5000.0   36 months   10.49%       162.49     B        B3   
3          24000.0   60 months   11.05%       522.42     B        B4   
4          14000.0   60 months   13.59%       322.79     C        C2   

        emp_title emp_length home_ownership  annual_inc verification_status  \
0         cashier  10+ years       MORTGAGE     48000.0     Source Verified   
1             ABM    6 years           RENT     57000.0     Source Verified   
2          driver  10+ years       MORTGAGE     55000.0        Not Verified   
3             NaN        NaN       MORTGAGE     43500.0        Not Verified   
4  Shipping Clerk  10+ years       MORTGAGE     48000.0     Source Verified   

  issue_d loan_status pymnt_plan  \
0  Mar-20     Current          n   
1  Feb-16     Current          n   
2  Jan-15  Fully Paid          n   
3  Jun-18     Current          n   
4  Jan-18  Fully Paid          n   

                                                 url             purpose  \
0  https://lendingclub.com/browse/loanDetail.acti...    home_improvement   
1  https://lendingclub.com/browse/loanDetail.acti...         credit_card   
2  https://lendingclub.com/browse/loanDetail.acti...         credit_card   
3  https://lendingclub.com/browse/loanDetail.acti...  debt_consolidation   
4  https://lendingclub.com/browse/loanDetail.acti...  debt_consolidation   

                     title zip_code addr_state    dti  delinq_2yrs  \
0         Home improvement    115xx         NY  20.25            1   
1  Credit card refinancing    708xx         LA  25.33            0   
2  Credit card refinancing    347xx         FL   7.22            0   
3       Debt consolidation    443xx         OH  24.55            0   
4       Debt consolidation    681xx         NE  24.30            0   

  earliest_cr_line  fico_range_low  fico_range_high  inq_last_6mths  \
0           Dec-04             665              669               0   
1           Jan-04             680              684               0   
2           Apr-10             665              669               0   
3           Oct-05             750              754               0   
4           Apr-08             735              739               1   

   mths_since_last_delinq  mths_since_last_record  open_acc  pub_rec  \
0                    19.0                   110.0         4        1   
1                     NaN                     NaN        10        0   
2                     NaN                    61.0         7        1   
3                     NaN                     NaN        15        0   
4                     NaN                     NaN        16        0   

   revol_bal revol_util  total_acc initial_list_status  out_prncp  \
0       1988     82.80%         12                   w    3816.34   
1      22697     64.50%         24                   w    4333.29   
2       6001     52.60%          9                   f       0.00   
3       6573     12.10%         23                   w   16317.69   
4      11894     27.10%         23                   w       0.00   

   out_prncp_inv   total_pymnt  total_pymnt_inv  total_rec_prncp  \
0        3816.34    266.950000           266.95           183.66   
1        4333.29  25496.160000         25496.16         19666.71   
2           0.00   5846.672836          5846.67          5000.00   
3       16317.69  12000.930000         12000.93         

In [11]:
# Copies the columns and descriptions from the data dictionary into a data frame for future recall.
# Initialize empty lists for LoanStatNew and Description
loanstatnew = []
description = []

# Iterate through each row in the DataFrame and populate lists
for _, row in df_data_dict.iterrows():
    loanstatnew.append(row['LoanStatNew'])
    description.append(row['Description'])

# Apply left-aligned styling to both headers and data cells
styled_df_data_dict = df_data_dict.style.set_properties(
    **{'text-align': 'left', 'white-space': 'nowrap'}
).set_table_styles(
    [{'selector': 'th', 'props': [('text-align', 'left')]}]
)

# Display styled DataFrame
styled_df_data_dict


### 01 Data Preparation and Cleaning

Perform thorough data cleaning on the provided dataset, including but not limited to the following steps:

* Handling missing values (imputation or removal)
* Converting data types to appropriate formats
* Removing duplicate records
* Detecting and handling outliers
* Standardizing and normalizing data
* Encoding categorical variables
* Cleaning and preprocessing string data
* Extracting features from date columns

Students are encouraged to perform additional data cleaning steps beyond those implemented in class (e.g., cleaning of additional columns).

#### Handling missing values (imputation or removal)

#### Converting data types to appropriate formats

#### Removing duplicate records

#### Detecting and handling outliers

#### Standardising and normalising data

#### Encoding categorical variables

#### Cleaning and preprocessing string data

#### Extracting features from date columns

## 02 Exploratory Data Analysis

Conduct an in-depth analysis of the dataset with a focus on the target variable. The analysis should include:

* Exploring the distribution, symmetry, and potential issues with the target variable.
* Using visualization techniques (e.g., histograms, box plots, scatter plots) and statistical analysis to explore relationships between the target variable and independent variables.
* Identifying important variables with predictive relevance.
* Determining which variables or levels can be excluded.
* Identifying variables with outliers and applying transformations if necessary.
* Handling missing values and explaining the chosen treatment.
* Examining interrelationships between independent variables and considering transformations.
* Assessing class balance and addressing any imbalance if needed.
* Summarizing insights and plans to leverage the information.

#### Exploring the Distribution, Symmetry, and Potential Issues with the Target Variable

#### Using Visualization Techniques to Explore Relationships Between the Target Variable and Independent Variables

#### Identifying Important Variables with Predictive Relevance

#### Determining Which Variables or Levels Can Be Excluded

#### Identifying Variables with Outliers and Applying Transformations if Necessary

#### Handling Missing Values and Explaining the Chosen Treatment

#### Examining Interrelationships Between Independent Variables and Considering Transformations

#### Assessing Class Balance and Addressing Any Imbalance if Needed

#### Summarizing Insights and Plans to Leverage the Information

### 03 Modelling

Recommend and justify a model to predict class membership of loan applications. The modeling phase should include:

* Selecting a baseline model for comparison.
* Recommending a challenger model with a detailed justification.
* Describing all data preprocessing steps and measurement of accuracy.
* Choosing appropriate models and evaluation metrics.
* Explaining the choice of models, preprocessing methods, and accuracy metrics.

### 04 Optional - Real-time scoring application

Build a "real-time" application that can score new loan application observations. The implementation details are at the consultant's discretion.